File: src/domain/repositories/order-ticket-repository.ts
import type { Repository } from "./repository.js";
import type { OrderTicket } from "../models/order-ticket.js";
import type { ID } from "../types/id.js";
import type { PaymentMethod } from "../types/payment-method.js";

/**
 * 注文チケットリポジトリのインターフェース
 * 基本的なCRUD操作に加え、チケット特有の検索・更新機能を提供します
 */
export interface OrderTicketRepository extends Repository<OrderTicket> {
    /**
     * チケット番号による検索
     * @param ticketNumber チケット番号
     */
    findByTicketNumber(ticketNumber: string): Promise<OrderTicket | null>;

    /**
     * 注文IDによる検索
     * @param orderId 注文ID
     */
    findByOrderId(orderId: ID): Promise<OrderTicket | null>;

    /**
     * 支払い方法による検索
     * @param paymentMethod 支払い方法
     */
    findByPaymentMethod(paymentMethod: PaymentMethod): Promise<OrderTicket[]>;

    /**
     * 支払い状態による検索
     * @param isPaid 支払い済みかどうか
     */
    findByPaymentStatus(isPaid: boolean): Promise<OrderTicket[]>;

    /**
     * 商品引渡し状態による検索
     * @param isDelivered 引渡し済みかどうか
     */
    findByDeliveryStatus(isDelivered: boolean): Promise<OrderTicket[]>;

    /**
     * 支払い状態を更新
     * @param id チケットID
     * @param isPaid 支払い状態
     * @param transactionId 決済トランザクションID（電子決済の場合）
     */
    updatePaymentStatus(
        id: ID,
        isPaid: boolean,
        transactionId?: string
    ): Promise<OrderTicket>;

    /**
     * 商品引渡し状態を更新
     * @param id チケットID
     * @param isDelivered 引渡し状態
     */
    updateDeliveryStatus(id: ID, isDelivered: boolean): Promise<OrderTicket>;

    /**
     * 特定の時間範囲のチケットを取得
     * @param startTime 開始時間
     * @param endTime 終了時間
     */
    findByTimeRange(startTime: Date, endTime: Date): Promise<OrderTicket[]>;

    /**
     * チケットの集計情報を取得
     * @param date 集計日
     */
    getDailySummary(date: Date): Promise<{
        total: number;
        byPaymentMethod: Record<PaymentMethod, number>;
        paid: number;
        delivered: number;
    }>;
}

File: internal/infrastructure/repositories/order_ticket_repository.go
package repositories

import (
	"context"

	"github.com/SeikoStudentCouncil/timeseats-backend/internal/domain/models"
	"github.com/SeikoStudentCouncil/timeseats-backend/internal/domain/repositories"
	"github.com/SeikoStudentCouncil/timeseats-backend/internal/domain/types"
	"gorm.io/gorm"
)

type orderTicketRepository struct {
	db *gorm.DB
}

func NewOrderTicketRepository(db *gorm.DB) repositories.OrderTicketRepository {
	return &orderTicketRepository{db: db}
}

func (r *orderTicketRepository) Create(ctx context.Context, ticket *models.OrderTicket) error {
	if err := r.db.WithContext(ctx).Create(ticket).Error; err != nil {
		return &repositories.RepositoryError{
			Operation: "Create",
			Err:       err,
		}
	}
	return nil
}

func (r *orderTicketRepository) FindByID(ctx context.Context, id types.ID) (*models.OrderTicket, error) {
	var ticket models.OrderTicket
	if err := r.db.WithContext(ctx).
		Preload("Order").
		Preload("Order.Items").
		Preload("Order.SalesSlot").
		First(&ticket, "id = ?", id).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, repositories.NewErrNotFound("OrderTicket", id)
		}
		return nil, &repositories.RepositoryError{
			Operation: "FindByID",
			Err:       err,
		}
	}
	return &ticket, nil
}

func (r *orderTicketRepository) FindAll(ctx context.Context) ([]models.OrderTicket, error) {
	var tickets []models.OrderTicket
	if err := r.db.WithContext(ctx).
		Preload("Order").
		Preload("Order.Items").
		Preload("Order.SalesSlot").
		Find(&tickets).Error; err != nil {
		return nil, &repositories.RepositoryError{
			Operation: "FindAll",
			Err:       err,
		}
	}
	return tickets, nil
}

func (r *orderTicketRepository) Update(ctx context.Context, ticket *models.OrderTicket) error {
	if err := r.db.WithContext(ctx).Save(ticket).Error; err != nil {
		return &repositories.RepositoryError{
			Operation: "Update",
			Err:       err,
		}
	}
	return nil
}

func (r *orderTicketRepository) Delete(ctx context.Context, id types.ID) error {
	result := r.db.WithContext(ctx).Delete(&models.OrderTicket{}, "id = ?", id)
	if result.Error != nil {
		return &repositories.RepositoryError{
			Operation: "Delete",
			Err:       result.Error,
		}
	}
	if result.RowsAffected == 0 {
		return repositories.NewErrNotFound("OrderTicket", id)
	}
	return nil
}

func (r *orderTicketRepository) FindByTicketNumber(ctx context.Context, ticketNumber string) (*models.OrderTicket, error) {
	var ticket models.OrderTicket
	if err := r.db.WithContext(ctx).
		Preload("Order").
		Preload("Order.Items").
		Preload("Order.SalesSlot").
		Where("ticket_number = ?", ticketNumber).
		First(&ticket).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, &repositories.RepositoryError{
				Operation: "FindByTicketNumber",
				Err:       err,
			}
		}
		return nil, &repositories.RepositoryError{
			Operation: "FindByTicketNumber",
			Err:       err,
		}
	}
	return &ticket, nil
}

func (r *orderTicketRepository) FindByOrderID(ctx context.Context, orderID types.ID) (*models.OrderTicket, error) {
	var ticket models.OrderTicket
	if err := r.db.WithContext(ctx).
		Preload("Order").
		Preload("Order.Items").
		Preload("Order.SalesSlot").
		Where("order_id = ?", orderID).
		First(&ticket).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, repositories.NewErrNotFound("OrderTicket", "")
		}
		return nil, &repositories.RepositoryError{
			Operation: "FindByOrderID",
			Err:       err,
		}
	}
	return &ticket, nil
}

func (r *orderTicketRepository) UpdatePaymentStatus(ctx context.Context, id types.ID, isPaid bool, transactionID *string) error {
	updates := map[string]interface{}{
		"is_paid": isPaid,
	}
	if transactionID != nil {
		updates["transaction_id"] = transactionID
	}

	result := r.db.WithContext(ctx).Model(&models.OrderTicket{}).
		Where("id = ?", id).
		Updates(updates)

	if result.Error != nil {
		return &repositories.RepositoryError{
			Operation: "UpdatePaymentStatus",
			Err:       result.Error,
		}
	}
	if result.RowsAffected == 0 {
		return repositories.NewErrNotFound("OrderTicket", id)
	}
	return nil
}

func (r *orderTicketRepository) UpdateDeliveryStatus(ctx context.Context, id types.ID, isDelivered bool) error {
	result := r.db.WithContext(ctx).Model(&models.OrderTicket{}).
		Where("id = ?", id).
		Update("is_delivered", isDelivered)

	if result.Error != nil {
		return &repositories.RepositoryError{
			Operation: "UpdateDeliveryStatus",
			Err:       result.Error,
		}
	}
	if result.RowsAffected == 0 {
		return repositories.NewErrNotFound("OrderTicket", id)
	}
	return nil
}
